### PR TITLE
Don't lint `multiple_inherent_impl` with generic arguments

### PR DESCRIPTION
**Pull request by Jarcho**
fixes: #5772
changelog: Treat different generic arguments as different types in `multiple_inherent_impl`

### CODE REVIEW

**Comment by camsteffen** [comment]
I don't think we should totally skip types with generic arguments. Instead we should detect if two impls have the exact same generic arguments. Since this is a restriction lint, some false positives are okay and even preferable to false negatives IMO.

**Comment by Jarcho** [comment]
I'll keep this open while i look into it.

**Comment by Jarcho** [comment]
That should do it. I'm not sure if that's the right way to add a dependency on `indexmap`.

File: clippy_lints/src/inherent_impl.rs [hunk 1]
```diff
@@ -43,7 +44,7 @@ declare_clippy_lint! {
 #[allow(clippy::module_name_repetitions)]
 #[derive(Default)]
 pub struct MultipleInherentImpl {
-    impls: DefIdMap<Span>,
+    impls: Vec<(LocalDefId, Span)>,
```
Comment (camsteffen) [inline clippy_lints/src/inherent_impl.rs:47]: Why not put the IndexMap right here?

Comment (Jarcho) [reply clippy_lints/src/inherent_impl.rs:47]: Lifetime issues. Lints need a `'static` lifetime, but the `IndexMap` uses `Ty<'_>` as the key. Technically those could be cast from a pointer to a usize.

Comment (camsteffen) [reply clippy_lints/src/inherent_impl.rs:47]: You could remove `Span`s by using `hir.span(hir.local_def_id_to_hir_id(impl_did))`.

Comment (camsteffen) [reply clippy_lints/src/inherent_impl.rs:47]: > Lifetime issues.

Ah ok.

Comment (camsteffen) [reply clippy_lints/src/inherent_impl.rs:47]: Any interest in removing the Spans? If not I'm fine merging as is.

File: clippy_lints/src/inherent_impl.rs [hunk 2]
```diff
@@ -56,33 +57,41 @@ impl<'tcx> LateLintPass<'tcx> for MultipleInherentImpl {
...
-                if let Some(initial_span) = impl_spans.next() {
-                    impl_spans.for_each(|additional_span| {
+            // Split the collected impls by type
+            let mut type_map = FxIndexMap::default();
+            for (ty, span) in self.impls.iter().map(|&(did, span)| (cx.tcx.type_of(did), span)) {
+                match type_map.entry(ty) {
+                    Entry::Vacant(e) => {
+                        e.insert(vec![span]);
+                    },
+                    Entry::Occupied(mut e) => e.get_mut().push(span),
```
Comment (camsteffen) [inline clippy_lints/src/inherent_impl.rs:78]: Thought of one more suggestion if you like. You could lint right here instead of adding to the map. Then it could be a HashMap with only the first impl per type.

File: clippy_lints/src/inherent_impl.rs [hunk 3]
```diff
@@ -56,33 +57,41 @@ impl<'tcx> LateLintPass<'tcx> for MultipleInherentImpl {
...
-                self.impls.insert(item.def_id.to_def_id(), item.span);
+                self.impls.push((item.def_id, item.span));
             }
         }
     }
 
     fn check_crate_post(&mut self, cx: &LateContext<'tcx>, krate: &'tcx Crate<'_>) {
         if !krate.items.is_empty() {
-            // Retrieve all inherent implementations from the crate, grouped by type
-            for impls in cx.tcx.crate_inherent_impls(def_id::LOCAL_CRATE).inherent_impls.values() {
```
Comment (camsteffen) [inline clippy_lints/src/inherent_impl.rs:71]: Could we continue to use `crate_inherent_impls` and avoid *any* data on the lint pass?

File: clippy_lints/src/inherent_impl.rs (hunk 2 above)
Comment (Jarcho) [reply clippy_lints/src/inherent_impl.rs:78]: That would be better. The only difference would be the order things are printed. It's currently sorted by type, then declaration order. This would make it solely by declaration order.

Comment (camsteffen) [reply clippy_lints/src/inherent_impl.rs:78]: That okay.

File: clippy_lints/src/inherent_impl.rs (hunk 3 above)
Comment (Jarcho) [reply clippy_lints/src/inherent_impl.rs:71]: Both this change, and the change with storing the span have the downside of adding a table lookup for each impl block. This would avoid the vec allocation, which might make it better though. At the very least it's a little simpler.

Comment (camsteffen) [reply clippy_lints/src/inherent_impl.rs:71]: You would only need to lookup spans when you lint. Is there another query I'm missing? Generally I think it's better to not store what can be queried.

Comment (Jarcho) [reply clippy_lints/src/inherent_impl.rs:71]: There's a macro check and a check for generic parameters. Both could be held off on until the second matching type is seen, so not every impl block.

Comment (camsteffen) [reply clippy_lints/src/inherent_impl.rs:71]: > There's a macro check and a check for generic parameters.

Oh yeah I forgot that. Whatever you decide then.

File: clippy_lints/src/lib.rs [hunk 4]
```diff
@@ -20,6 +20,7 @@
 
 // FIXME: switch to something more ergonomic here, once available.
 // (Currently there is no way to opt into sysroot crates without `extern crate`.)
+extern crate indexmap;
```
Comment (flip1995) [inline clippy_lints/src/lib.rs:23]: Adding `indexmap` as a dep should be ok, since rustc also uses this dep, so we shouldn't see conflicts with rustc.

But what is the reason to use indexmap from the sysroot rather than a dep? (Or formulated a different way: Why does rustc-dev contain indexmap?)

Comment (Jarcho) [reply clippy_lints/src/lib.rs:23]: `rustc_data_structures` exports `FxIndexMap`, but not it's matching entry type.

File: clippy_lints/src/inherent_impl.rs (hunk 3 above)
Comment (Jarcho) [reply clippy_lints/src/inherent_impl.rs:71]: Tried switching to `crate_inherent_impls`, but it returns a hash map from the type definition to it's impl blocks. This causes the lint output to be in an undefined order which is both a small downgrade for people reading it, and can cause the tests to break for no reason.

File: clippy_lints/src/lib.rs (hunk 4 above)
Comment (Jarcho) [reply clippy_lints/src/lib.rs:23]: I was using `Entry` with `FxIndexMap` which requires using the same `indexmap` dependency as rustc does. Not a problem anymore as the order isn't needed anymore.

File: clippy_lints/src/inherent_impl.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_lints/src/inherent_impl.rs:47]: Is there really much improvement from doing that? It's just changing the small amount of memory use for having to lookup the span later in the HIR map.

**Review by camsteffen — CHANGES_REQUESTED** [review]
Looking good! Just need some comments as it's a little hard to parse.

**Comment by camsteffen** [comment]
Please squash commits and r=me
